src/System.Private.Xml/src/System/Xml/Dom/XmlDocument.cs

```diff
@@ -94,28 +145,51 @@ protected internal XmlDocument(XmlImplementation imp) : base()
             _implementation = imp;
             _domNameTable = new DomNameTable(this);
 
-            // force the following string instances to be default in the nametable
-            XmlNameTable nt = this.NameTable;
-            nt.Add(string.Empty);
-            strDocumentName = nt.Add("#document");
-            strDocumentFragmentName = nt.Add("#document-fragment");
-            strCommentName = nt.Add("#comment");
-            strTextName = nt.Add("#text");
-            strCDataSectionName = nt.Add("#cdata-section");
-            strEntityName = nt.Add("#entity");
-            strID = nt.Add("id");
-            strNonSignificantWhitespaceName = nt.Add("#whitespace");
-            strSignificantWhitespaceName = nt.Add("#significant-whitespace");
-            strXmlns = nt.Add("xmlns");
-            strXml = nt.Add("xml");
-            strSpace = nt.Add("space");
-            strLang = nt.Add("lang");
-            strReservedXmlns = nt.Add(XmlReservedNs.NsXmlNs);
-            strReservedXml = nt.Add(XmlReservedNs.NsXml);
-            strEmpty = nt.Add(String.Empty);
+            strXmlns = Xmlns;
+            strXml = Xml;
+            strReservedXmlns = XmlReservedNs.NsXmlNs;
+            strReservedXml = XmlReservedNs.NsXml;
             baseURI = String.Empty;
-
             objLock = new object();
+
+            if (imp.NameTable is NameTable nt)
```

Comment:
this may not work correctly for derived classes, should be: `imp.NameTable.GetType() == typeof(NameTable)`

Reply:
Yes, you're right. I fail to see an useful implementation deriving from NameTable that can have trouble with that but perhaps I'm not creative enough for that :smile: fixing it...